File: cpp/backend/snapshot_test.cc
#include "backend/snapshot.h"

#include <algorithm>
#include <cstddef>
#include <span>
#include <string>
#include <vector>

#include "absl/status/status.h"
#include "absl/status/statusor.h"
#include "common/hash.h"
#include "common/status_test_util.h"
#include "gmock/gmock.h"
#include "gtest/gtest.h"

namespace carmen::backend {
namespace {

// TestProof is an example proof implementation for test cases in this file.
using TestProof = Hash;
static_assert(Proof<TestProof>);

// TestPart is an example part implementation for tests cases in this file.
class TestPart {
 public:
  using Proof = TestProof;

  TestPart(Proof proof, std::span<const std::byte> data)
      : proof_(proof), data_(data.begin(), data.end()) {}

  static absl::StatusOr<TestPart> FromBytes(std::span<const std::byte> data) {
    if (data.size() < sizeof(Proof)) {
      return absl::InvalidArgumentError(
          "Invalid encoding of TestPart, too few bytes");
    }
    Proof proof;
    proof.SetBytes(data.subspan(0, sizeof(Proof)));
    return TestPart(proof, data.subspan(sizeof(Proof)));
  }

  const Proof& GetProof() const { return proof_; }

  bool Verify() const { return GetSha256Hash(data_) == proof_; }

  const std::vector<std::byte>& GetData() const { return data_; }

  std::vector<std::byte> ToBytes() const {
    // Serialized as proof, followed by data.
    std::vector<std::byte> res;
    res.reserve(sizeof(Hash) + data_.size());
    std::span<const std::byte> hash = proof_;
    res.insert(res.end(), hash.begin(), hash.end());
    res.insert(res.end(), data_.begin(), data_.end());
    return res;
  }

 private:
  Proof proof_;
  std::vector<std::byte> data_;
};

static_assert(Part<TestPart>);

// TestSnapshot is an example snapshot implementation demonstrating its concept
// and serving as a example implementation for the tests in this file.
class TestSnapshot {
 public:
  static constexpr std::size_t part_size = 4;
  using Proof = TestProof;
  using Part = TestPart;

  TestSnapshot(Proof proof, std::span<const std::byte> data)
      : proof_(proof), data_(data.begin(), data.end()) {
    // Some extra padding to make the internal snapshot data size a multiple of
    // the data size.
    auto extra = data_.size() % part_size;
    if (extra > 0) {
      data_.resize(data_.size() + (part_size - extra));
    }
    assert(data_.size() >= data.size());
    assert(data_.size() < data.size() + part_size);
    assert(data_.size() % part_size == 0);

    // In a real setup, those part hashes would come from another source.
    for (std::size_t i = 0; i < GetSize(); i++) {
      proofs_.push_back(
          GetSha256Hash(std::span(data_).subspan(i * part_size, part_size)));
    }
  }

  std::size_t GetSize() const {
    return data_.size() / part_size + (data_.size() % part_size > 0);
  }

  absl::StatusOr<Part> GetPart(std::size_t i) const {
    if (i >= GetSize()) {
      return absl::NotFoundError("no such part");
    }
    auto data = std::span(data_).subspan(i * part_size, part_size);
    return Part(GetSha256Hash(data), data);
  }

  Proof GetProof() const { return proof_; }

  absl::StatusOr<Proof> GetProof(std::size_t i) const {
    return i < proofs_.size() ? proofs_[i] : Proof{};
  }

  absl::Status VerifyProofs() const {
    Sha256Hasher hasher;
    for (const auto& proof : proofs_) {
      hasher.Ingest(proof);
    }
    Proof should = hasher.GetHash();
    if (should != proof_) {
      return absl::InternalError("Proofs are not consistent");
    }
    return absl::OkStatus();
  }

 private:
  // The full proof of the snapshot.
  Proof proof_;

  // The proofs of the individual parts.
  std::vector<Proof> proofs_;

  // The raw data of the snapshot, padded with 0s have a size that is a multiple
  // of the part_size.
  std::vector<std::byte> data_;
};

static_assert(Snapshot<TestSnapshot>);

class TestData {
 public:
  using Snapshot = TestSnapshot;

  TestData(std::string_view str) : data_(str) {}

  static absl::StatusOr<TestData> Restore(const Snapshot& snapshot) {
    std::string data;
    for (std::size_t i = 0; i < snapshot.GetSize(); i++) {
      ASSIGN_OR_RETURN(auto part, snapshot.GetPart(i));
      auto substr = part.GetData();
      data.append(reinterpret_cast<const char*>(substr.data()), substr.size());
    }
    // Remove the extra padding that is stored in the snapshot.
    while (data.size() > 0 && data[data.size() - 1] == '\0') {
      data.resize(data.size() - 1);
    }
    TestData res("");
    res.data_ = std::move(data);
    return res;
  }

  absl::StatusOr<TestProof> GetProof() const {
    std::string padded = data_;
    while (padded.size() % 4 != 0) {
      padded += '\0';
    }
    Sha256Hasher global_hasher;
    for (std::size_t i = 0; i < padded.size(); i += 4) {
      Sha256Hasher part_hasher;
      part_hasher.Ingest(std::as_bytes(std::span(padded).subspan(i, 4)));
      global_hasher.Ingest(part_hasher.GetHash());
    }
    return global_hasher.GetHash();
  }

  absl::StatusOr<Snapshot> CreateSnapshot() const {
    ASSIGN_OR_RETURN(auto proof, GetProof());
    return Snapshot(proof, std::as_bytes(std::span(data_)));
  }

  TestData& operator=(std::string_view data) {
    data_ = data;
    return *this;
  }

  std::string& GetData() { return data_; }

 private:
  std::string data_;
};

static_assert(Snapshotable<TestData>);

TEST(Snapshot, SnapshotCanBeCreated) {
  TestData data("some test data");
  ASSERT_OK_AND_ASSIGN(auto snapshot, data.CreateSnapshot());
  ASSERT_THAT(data.GetData().size(), 14);
  EXPECT_THAT(snapshot.GetSize(), 14 / 4 + 1);
}

TEST(Snapshot, ProofOfDataEqualsProofOfSnapshot) {
  TestData data("some test data");
  ASSERT_OK_AND_ASSIGN(auto data_proof, data.GetProof());

  ASSERT_OK_AND_ASSIGN(auto snapshot, data.CreateSnapshot());
  auto shot_proof = snapshot.GetProof();

  EXPECT_EQ(data_proof, shot_proof);
}

TEST(Snapshot, ChangingTheDataDoesNotChangeTheSnapshotProof) {
  TestData data("some test data");
  ASSERT_OK_AND_ASSIGN(auto old_data_proof, data.GetProof());

  ASSERT_OK_AND_ASSIGN(auto snapshot, data.CreateSnapshot());
  auto old_shot_proof = snapshot.GetProof();

  data = "some other content";

  // The proof of the data has changed.
  ASSERT_OK_AND_ASSIGN(auto new_data_proof, data.GetProof());
  EXPECT_NE(old_data_proof, new_data_proof);

  // The proof of the snapshot has not changed.
  auto new_shot_proof = snapshot.GetProof();
  EXPECT_EQ(old_shot_proof, new_shot_proof);
}

TEST(Snapshot, SnapshotCanBeRestored) {
  TestData data("some test data");
  ASSERT_OK_AND_ASSIGN(auto data_proof, data.GetProof());
  ASSERT_OK_AND_ASSIGN(auto snapshot, data.CreateSnapshot());
  data = "some other content";

  ASSERT_OK_AND_ASSIGN(auto restored, TestData::Restore(snapshot));
  EXPECT_EQ(restored.GetData(), "some test data");
  EXPECT_THAT(restored.GetProof(), data_proof);
}

TEST(Snapshot, PartProofsCanBeVerified) {
  TestData data("some test data");
  ASSERT_OK_AND_ASSIGN(auto snapshot, data.CreateSnapshot());
  ASSERT_LT(1, snapshot.GetSize());
  for (std::size_t i = 0; i < snapshot.GetSize(); i++) {
    ASSERT_OK_AND_ASSIGN(auto part, snapshot.GetPart(i));
    EXPECT_THAT(snapshot.GetProof(i), part.GetProof());
  }
}

TEST(Snapshot, SnapshotProofsCanBeVerified) {
  TestData data("some test data");
  ASSERT_OK_AND_ASSIGN(auto snapshot, data.CreateSnapshot());
  ASSERT_LT(1, snapshot.GetSize());
  EXPECT_OK(snapshot.VerifyProofs());
}

}  // namespace
}  // namespace carmen::backend

File: go/backend/snapshot_test.go
package backend_test

import (
	"crypto/sha256"
	"fmt"
	"testing"

	"github.com/Fantom-foundation/Carmen/go/backend"
	"github.com/Fantom-foundation/Carmen/go/common"
)

// This test file contains an example implementation of a data structure
// (myDataStructure) supporting snapshots. The snapshot type is mySnapshot,
// the parts are of type myPart, and the proofs are of type myProof.

// ---------------------------------- Proof -----------------------------------

type myProof struct {
	hash common.Hash
}

func (p *myProof) Equal(other backend.Proof) bool {
	trg, ok := other.(*myProof)
	return ok && trg.hash == p.hash
}

func (p *myProof) ToBytes() []byte {
	return p.hash[:]
}

func getProofFor(data []byte) myProof {
	hash := sha256.New()
	hash.Write(data)
	var res common.Hash
	hash.Sum(res[:])
	return myProof{res}
}

// ----------------------------------- Part -----------------------------------

type myPart struct {
	proof *myProof
	data  []byte
}

func (p *myPart) GetProof() backend.Proof {
	return p.proof
}

func (p *myPart) Verify() bool {
	have := getProofFor(p.data)
	return p.GetProof().Equal(&have)
}

func (p *myPart) ToBytes() []byte {
	res := p.proof.ToBytes()
	res = append(res, p.data...)
	return res
}

// --------------------------------- Snapshot ---------------------------------

type mySnapshot struct {
	proof  *myProof
	size   int
	source mySnapshotDataSource
}

func (e *mySnapshot) GetRootProof() backend.Proof {
	return e.proof
}

func (e *mySnapshot) GetNumParts() int {
	return e.size
}

func (e *mySnapshot) GetProof(part_number int) (backend.Proof, error) {
	if part_number >= e.size {
		return nil, fmt.Errorf("no such part")
	}
	return e.source.GetProof(part_number)
}

func (e *mySnapshot) GetPart(part_number int) (backend.Part, error) {
	proof, err := e.GetProof(part_number)
	if err != nil {
		return nil, err
	}
	data, err := e.source.GetData(part_number)
	if err != nil {
		return nil, err
	}
	return &myPart{proof: proof.(*myProof), data: data}, nil
}

func (e *mySnapshot) VerifyRootProof() error {
	h := sha256.New()
	for i := 0; i < e.GetNumParts(); i++ {
		proof, err := e.GetProof(i)
		if err != nil {
			return err
		}
		h.Write(proof.(*myProof).ToBytes())
	}
	var hash common.Hash
	h.Sum(hash[:])
	if hash != e.proof.hash {
		return fmt.Errorf("proof validation failed")
	}
	return nil
}

func (e *mySnapshot) GetData() backend.SnapshotData {
	return e
}

func (e *mySnapshot) GetMetaData() ([]byte, error) {
	res := []byte{}
	res = append(res, byte(e.GetNumParts()))
	res = append(res, e.proof.hash[:]...)
	return res, nil
}

func (e *mySnapshot) GetProofData(part_number int) ([]byte, error) {
	proof, err := e.GetProof(part_number)
	if err != nil {
		return nil, err
	}
	return proof.ToBytes(), nil
}

func (e *mySnapshot) GetPartData(part_number int) ([]byte, error) {
	part, err := e.GetPart(part_number)
	if err != nil {
		return nil, err
	}
	return part.ToBytes(), nil
}

func (e *mySnapshot) Release() error {
	return nil // nothing to do
}

func createMySnapshotFromData(data backend.SnapshotData) (*mySnapshot, error) {
	metadata, err := data.GetMetaData()
	if err != nil {
		return nil, err
	}
	if len(metadata) != common.HashSize+1 {
		return nil, fmt.Errorf("invalid meta data encoding, invalid number of bytes")
	}
	size := int(metadata[0])
	var hash common.Hash
	copy(hash[:], metadata[1:])
	return &mySnapshot{&myProof{hash}, size, &SnapshotDataSource{data}}, nil
}

type mySnapshotDataSource interface {
	GetProof(part_number int) (*myProof, error)
	GetData(part_number int) ([]byte, error)
}

type SnapshotDataSource struct {
	snapshot backend.SnapshotData
}

func (s *SnapshotDataSource) GetProof(part_number int) (*myProof, error) {
	data, err := s.snapshot.GetProofData(part_number)
	if err != nil {
		return nil, err
	}
	if len(data) != common.HashSize {
		return nil, fmt.Errorf("invalid length of example proof")
	}
	var hash common.Hash
	copy(hash[:], data)
	return &myProof{hash}, nil
}

func (s *SnapshotDataSource) GetData(part_number int) ([]byte, error) {
	data, err := s.snapshot.GetPartData(part_number)
	if err != nil {
		return nil, err
	}
	if len(data) < common.HashSize {
		return nil, fmt.Errorf("invalid length of example part")
	}
	return data[common.HashSize:], nil
}

// ------------------------------- DataStructure ------------------------------

type myDataStructure struct {
	proofs []myProof
	data   [][]byte
}

func (e *myDataStructure) Set(pos int, data []byte) {
	for len(e.data) <= pos {
		e.data = append(e.data, nil)
		e.proofs = append(e.proofs, getProofFor(nil))
	}
	e.data[pos] = make([]byte, len(data))
	copy(e.data[pos], data)
	e.proofs[pos] = getProofFor(data)
}

func (e *myDataStructure) Get(pos int) []byte {
	if pos < 0 || pos >= len(e.data) {
		return nil
	}
	return e.data[pos]
}

func (d *myDataStructure) GetProof() (backend.Proof, error) {
	h := sha256.New()
	for _, cur := range d.proofs {
		h.Write(cur.hash[:])
	}
	var hash common.Hash
	h.Sum(hash[:])
	return &myProof{hash}, nil
}

func (d *myDataStructure) CreateSnapshot() (backend.Snapshot, error) {
	proof, err := d.GetProof()
	if err != nil {
		return nil, err
	}
	proofs := make([]myProof, len(d.proofs))
	copy(proofs, d.proofs)
	data := make([][]byte, len(d.data))
	for i, cur := range d.data {
		cpy := make([]byte, len(cur))
		copy(cpy, cur)
		data[i] = cpy
	}
	return &mySnapshot{proof.(*myProof), len(data), &myDataStructureCopy{proofs, data}}, nil
}

func (d *myDataStructure) Restore(data backend.SnapshotData) error {
	snapshot, err := createMySnapshotFromData(data)
	if err != nil {
		return err
	}

	d.data = make([][]byte, snapshot.GetNumParts())

	for i := 0; i < snapshot.GetNumParts(); i++ {
		part, err := snapshot.GetPart(i)
		if err != nil {
			return err
		}
		cpy := make([]byte, len(part.(*myPart).data))
		copy(cpy, part.(*myPart).data)
		d.data[i] = cpy
	}
	return nil
}

type myDataStructureCopy struct {
	proofs []myProof
	data   [][]byte
}

func (s *myDataStructureCopy) GetProof(part_number int) (*myProof, error) {
	if part_number < 0 || part_number >= len(s.proofs) {
		return nil, fmt.Errorf("no such part")
	}
	return &s.proofs[part_number], nil
}

func (s *myDataStructureCopy) GetData(part_number int) ([]byte, error) {
	if part_number < 0 || part_number >= len(s.proofs) {
		return nil, fmt.Errorf("no such part")
	}
	return s.data[part_number], nil
}

// ----------------------------------- Tests ----------------------------------

func TestMyDataStructureImplementsInterfaces(t *testing.T) {
	var _ backend.Proof = &myProof{}
	var _ backend.Part = &myPart{}
	var _ backend.Snapshot = &mySnapshot{}
	var _ backend.Snapshotable = &myDataStructure{}
}

func TestSnapshotCanBeCreatedAndRestored(t *testing.T) {
	original := &myDataStructure{}
	original.Set(1, []byte{1, 2, 3})
	original.Set(2, []byte{4, 5})
	original.Set(3, []byte{7, 8, 9})

	snapshot, err := original.CreateSnapshot()
	if err != nil {
		t.Errorf("failed to create snapshot: %v", err)
		return
	}
	if snapshot == nil {
		t.Errorf("failed to create snapshot")
		return
	}

	// clear the original data structure, to eliminate the old copy.
	original.data = [][]byte{}

	recovered := &myDataStructure{}
	if err := recovered.Restore(snapshot.GetData()); err != nil {
		t.Errorf("failed to sync to snapshot: %v", err)
		return
	}

	common.AssertArraysEqual(t, recovered.Get(1), []byte{1, 2, 3})
	common.AssertArraysEqual(t, recovered.Get(2), []byte{4, 5})
	common.AssertArraysEqual(t, recovered.Get(3), []byte{7, 8, 9})

	if err := snapshot.Release(); err != nil {
		t.Errorf("failed to release snapshot: %v", err)
	}
}

func TestSnapshotCanBeCreatedAndValidated(t *testing.T) {
	structure := &myDataStructure{}
	structure.Set(1, []byte{1, 2, 3})
	structure.Set(2, []byte{4, 5})
	structure.Set(3, []byte{7, 8, 9})

	snapshot, err := structure.CreateSnapshot()
	if err != nil {
		t.Errorf("failed to create snapshot: %v", err)
		return
	}
	if snapshot == nil {
		t.Errorf("failed to create snapshot")
		return
	}

	remote, err := createMySnapshotFromData(snapshot.GetData())
	if err != nil {
		t.Fatalf("failed to create snapshot from snapshot data: %v", err)
	}

	// Test direct and serialized snapshot data access.
	for _, cur := range []backend.Snapshot{snapshot, remote} {

		// The root proof should be equivalent.
		want, err := structure.GetProof()
		if err != nil {
			t.Errorf("failed to get root proof from data structure")
		}

		have := cur.GetRootProof()
		if !want.Equal(have) {
			t.Errorf("root proof of snapshot does not match proof of data structure")
		}

		if err := cur.VerifyRootProof(); err != nil {
			t.Errorf("snapshot invalid, inconsistent proofs")
		}

		// Verify all pages
		for i := 0; i < cur.GetNumParts(); i++ {
			want, err := cur.GetProof(i)
			if err != nil {
				t.Errorf("failed to fetch proof of part %d", i)
			}
			part, err := cur.GetPart(i)
			if err != nil {
				t.Errorf("failed to fetch part %d", i)
			}
			if !want.Equal(part.GetProof()) {
				t.Errorf("proof of part does not equal proof provided by snapshot")
			}
			if !part.Verify() {
				t.Errorf("failed to verify content of part %d", i)
			}
		}
	}

	if err := remote.Release(); err != nil {
		t.Errorf("failed to release remote snapshot: %v", err)
	}
	if err := snapshot.Release(); err != nil {
		t.Errorf("failed to release snapshot: %v", err)
	}
}
